### PR TITLE
Immediate update of snippets

### LINKED ISSUE #707: snippets don't update when content of message/reply is updated
## STR

0. Submit a document so that there is a fresh submission to download in the next step.
1. Login to the client.
2. See the snippet displays our placeholder message.

## Expected behavior

As the message downloads, the content updates in the sidebar also.

## Actual behavior

The content is static, and will only refresh after a full sync (which will call the controller's `update_sources` method to update the sidebar). 

Screen Shot 2020-01-17 at 8 42 12 PM

## Comments

We do emit a signal each time a file, message, or reply is "ready" (i.e. has been successfully downloaded and decrypted), those signals are [here](https://github.com/freedomofpress/securedrop-client/blob/master/securedrop_client/logic.py#L131), so we could use make use of them in the source list view to update the source list view elements when the content is updated.

### PR DESCRIPTION
**Pull request by ntoll**
# Description

Ready for review.

Fixes #707.

This is a preview. **There are performance issues with the code as it stands** (tested with 200 sources), but I wanted to give you a sense of the progress made so far.

It took a while to work around SqlAlchemy, but got there in the end.

* On start, the snippets are updated as soon as the messages / replies etc are downloaded from the server.
* If a new message comes in while the client is active, the snippet is appropriately updated.
* When a new reply is sent and succeeds, the snippet is appropriately updated.
* When a file is downloaded from the server, the snippet is appropriately updated.

Screenie of some of this here:

snippet_updates

Performance refactoring currently underway.

# Test Plan

I've updated the unit tests. Please check with Eyeball mk.1 :eyes: for functionality described above.

Performance sanity check with 200 sources currently results in an unresponsive app while the messages / replies are downloading.

# Checklist

If these changes modify code paths involving cryptography, the opening of files in VMs or network (via the RPC service) traffic, Qubes testing in the staging environment is required. For fine tuning of the graphical user interface, testing in any environment in Qubes is required. Please check as applicable:

 - [ ] I have tested these changes in the appropriate Qubes environment
 - [ ] I do not have an appropriate Qubes OS workstation set up (the reviewer will need to test these changes)
 - [X] These changes should not need testing in Qubes

If these changes add or remove files other than client code, packaging logic (e.g., the AppArmor profile) may need to be updated. Please check as applicable:

 - [ ] I have submitted a separate PR to the [packaging repo](https://github.com/freedomofpress/securedrop-debian-packaging)
 - [X] No update to the packaging logic (e.g., AppArmor profile) is required for these changes
 - [ ] I don't know and would appreciate guidance

### CODE REVIEW

File: securedrop_client/gui/widgets.py [hunk 1]
```diff
@@ -1018,21 +1018,32 @@ def setup(self, controller):
         """
         self.controller = controller
         self.star.setup(self.controller)
+        self.controller.message_ready.connect(self.set_snippet)
+        self.controller.reply_ready.connect(self.set_snippet)
+        self.controller.reply_succeeded.connect(self.set_snippet)
+        self.controller.file_ready.connect(self.set_snippet)
```
Comment (creviera) [inline securedrop_client/gui/widgets.py]: headsup: you may need to connect one more signal if [this](https://github.com/freedomofpress/securedrop-client/pull/724/files#diff-16a175b4662541f3889b432ac9e7e0e8R1942) gets merged first

**Comment by ntoll** [comment]
I've managed to fix the unresponsive UI problem, fixed up the unit tests and tested with 200 sources to make sure it all works fine. Feedback most welcome..!

**Comment by rmol** [comment]
Looks good so far -- responsiveness is much improved, and the snippet is updating. I did notice a delay updating the message in the conversation pane: a box appeared for a new message but was empty, though the snippet under the source was updated with the message content ("This space intentionally left blank."):
image
This did self-correct after a few seconds, presumably at the next sync?

**Comment by rmol** [comment]
Replies can likewise stay blank for a bit after they're fetched and the snippet updated. Clicking on another source and back redraws them.

Downloading a file updated the snippet immediately, but the "downloading" spinner got stuck on, and this isn't cleared up by clicking another source then returning:

image

I should probably have mentioned that I'm testing on Qubes, under i3.

**Comment by rmol** [comment]
Same behavior observed under XFCE.

**Comment by ntoll** [comment]
@rmol aha... thank you. I'll take a look. I've been able to see this too. I suspect the blanks are an interesting intermediate state where the message is downloaded, but not yet decrypted. I'll need to check.

**Comment by ntoll** [comment]
I'll fix the downloading spinner in the morning.

**Comment by ntoll** [comment]
@creviera please confirm the spinner behaviour is not related to this branch (but is the problem you saw when testing the decryption story on the client side). If this is "just a bug" and not anything to do with this branch, I'll just ignore it and let it be fixed in another ticket.

**Comment by creviera** [comment]
@ntoll https://github.com/freedomofpress/securedrop-client/issues/754 only occurs when the decryption fails, so this is unrelated to this PR

**Comment by ntoll** [comment]
@creviera cool... so I won't fix in this PR. :-)

**Pull request merged**